docs/developers.md

```diff
@@ -20,80 +38,297 @@ For completed APB examples, take a look at some of the APBs in the [ansibleplayb
+
+* Provision
+```yaml
+- name: create deployment config
+  openshift_v1_deployment_config:
+    name: hello-world
+    namespace: '{{ namespace }}'
+    labels:
+      app: hello-world
+      service: hello-world
+    replicas: 1
+    selector:
+      app: hello-world
+      service: hello-world
+    spec_template_metadata_labels:
+      app: hello-world
+      service: hello-world
+    containers:
+    - env:
+      image: docker.io/ansibleplaybookbundle/hello-world:latest
+      name: hello-world
+      ports:
+      - container_port: 8080
+        protocol: TCP
+```
+
+* Deprovision
+```yaml
+- openshift_v1_deployment_config:
+    name: hello-world
+    namespace: '{{ namespace }}'
+    state: absent
+```
+
+## Route
+The following is an example of creating a route named `hello-world` which maps to service `hello-world`.
+* Provision
+```yaml
+- name: create hello-world route
+  openshift_v1_route:
+    name: hello-world
+    namespace: '{{ namespace }}'
+    spec_port_target_port: web
+    labels:
+      app: hello-world
+      service: hello-world
+    to_name: hello-world
+```
+
+* Deprovision
+```yaml
+- openshift_v1_route:
+    name: hello-world
+    namespace: '{{ namespace }}'
+    state: absent
+```
+
+## Persistent Volume
+The following is an example of creating a persistent volume claim resource and deployment config that uses it.
+* Provision
+```yaml
+# Persistent volume resource
+- name: create volume claim
+  k8s_v1_persistent_volume_claim:
+    name: hello-world-db
+    namespace: '{{ namespace }}'
+    state: present
+    access_modes:
+      - ReadWriteOnce
+    resources_requests:
+      storage: 1Gi
+
+
+# In addition to the resource, we need to add our volume to the deployment config declaration. 
+# The following is an example deployment config with a persistent volume.
+- name: create hello-world-db deployment config
+  openshift_v1_deployment_config:
+    name: hello-world-db
+    ---
+    volumes:
+    - name: hello-world-db
+      persistent_volume_claim:
+        claim_name: hello-world-db
+      test: false
+      triggers:
+      - type: ConfigChange
+```
+
+* Deprovision
+```yaml
+- openshift_v1_deployment_config:
+    name: hello-world-db
+    namespace: '{{ namespace }}'
+    state: absent
+
+- k8s_v1_persistent_volume_claim:
+    name: hello-world-db
+    namespace: '{{ namespace }}'
+    state: absent
+
+```
+
+# Tips and Tricks
+
+## Optional Variables
+You can add optional variables to an Ansible Playbook Bundle by using environment variables. To pass variables into an APB, you will need to escape the variable substitution in your `.yml` files. For example, the below is a section of the [main.yml](https://github.com/fusor/apb-examples/blob/master/etherpad-apb/roles/provision-etherpad-apb/tasks/main.yml#L89) in the [etherpad-apb](https://github.com/fusor/apb-examples/tree/master/etherpad-apb):
```

Comment:
"... the below is a section of the [main.yml]" should read "... the section below is of the [main.yml]"